modules/dns_dnscmd/dns_dnscmd_main.rb

```diff
@@ -37,8 +37,7 @@ def execute(cmd, msg = nil, error_only = false)
       tsecs = 5
       response = nil
       interpreter = Proxy::SETTINGS.x86_64 ? 'c:\windows\sysnative\cmd.exe' : 'c:\windows\system32\cmd.exe'
-      command = interpreter + ' /c c:\Windows\System32\dnscmd.exe ' + "#{@server} #{cmd}"
-
+      command = [interpreter, ' /c c:\Windows\System32\dnscmd.exe ', @server] + escape_for_cmdexe(cmd)
```

Comment:
I think each individual argument should be an entry in the array
```suggestion
      command = [interpreter, '/c', 'c:\Windows\System32\dnscmd.exe', @server] + escape_for_cmdexe(cmd)
```

It should also deal with `cmd` being an array. That way you can fix every call to `execute` to properly pass all arguments.

Reply:
I wouldn't mind having `/c` as another array item, however I don't think it's necessary to support also array for `cmd`. Given it is added to the command array, `popen3` take it as an argument and takes care of any special chars, it just won't format whitespaces and `cmd` can contain multiple spaces between args. The important thing is, it's added to the single command array.

Reply:
I don't think combining arguments in a long string is even valid. Just to be clear, I'd like to avoid using a shell at all. This would be the second form in https://docs.ruby-lang.org/en/2.4.0/Process.html#method-c-spawn. By avoiding a shell, there is no need to escape anything.

To me it's odd to combine multiple arguments in a string only to split it later. That's a step that can only introduce problems.

Reply:
Oh I think I understand where we misunderstood each other.

1) Yes, let's use splatted array form, meaning `cmdname, arg1, ...` because that performs some escaping and does not use shell.

2) The arg1 could contain more arguments, it could be `"/c c:\Windows\System32\dnscmd.exe"` and it AFAICT does not cause any harm. I agree it feels odd but is valid. I think that converting all existing command strings to arrays just because of it would result in very hard to read snippets. But it's not a strong opinion. The most important thing is to always have a command in the array form splitting at least the cmd name from the arguments.

Still, if you test the example from above and pass `'echo', 'hello', "& whatever` you'll see it cuts off everything after `&`. To avoid that, the characters need to be escaped. And that should be the content of `escape_for_cmdexe`, it should imho not be splitting strings by spaces.

Reply:
I want to avoid regressions, this code is poorly tested and I asked Anna only to do a simple approach by escaping cmd.exe special characters. Please let's not go into bigger refactoring, honestly, I was hoping to merge this patch without testing it as I have no idea how an array passed into `popen3` would work on Windows and I don't have Windows DNS setup available.

Reply:
I that case I withdraw myself as a reviewer. I don't see the point of making a change in the first place then.

Right now I'd argue there is already no real risk at all for a shell escape. This class has a well defined API and all the inputs are validated (in [dns_api](https://github.com/theforeman/smart-proxy/blob/develop/modules/dns/dns_api.rb)).

If we don't have a setup to validate this and there is no real risk for exploitation, we're only introducing risk for regression. I thought this was about implementing a proper solution.

I'll also leave a note that `escape_for_shell` in util right now is unused within this codebase, but it's only there compatibility for other modules. I'd strongly be in favor of deprecating it and eventually removing it, but this PR takes it in the opposite direction.

Reply:
Maybe I should have been more explicit, we have a setup to validate this and that's what I actually did. I ran it on Windows. The things I wrote above is a result on the actual test of the `popen3` behavior on Windows. Array aproach is good still needed some escaping. But I'm totally fine with fixing just part of it, still better than what we have now. Leaving up to @lzap to find the right scope of this PR.